x-pack/metricbeat/module/mssql/performance/_meta/data.json

```diff
@@ -0,0 +1,32 @@
+{
+    "@timestamp": "2017-10-12T08:05:34.853Z",
+    "agent": {
+        "hostname": "host.example.com",
+        "name": "host.example.com"
+    },
+    "event": {
+        "dataset": "mssql.performance",
+        "duration": 115000,
+        "module": "mssql"
+    },
+    "metricset": {
+        "name": "performance"
+    },
+    "mssql": {
+        "performance": {
+            "batch_requests": {},
```

Comment:
Probably needs an other generation?

Reply:
There's a problem, I don't see any obvious solution so some help will be appreciated. The problem is that each query generates a single event value so in `data.json` only the last query is visible (the other queries or events seems to be overriden).

I see two scenarios here: 

* Join all queries in a single query that produces all results. It must be somehow possible.
* Leave it like this, a single event for each query. Create `data.json` manually. Very fast and easy.

All queries are rather slow in my Docker environment. Just to run the integration tests takes `1.053s` for a total of 9 queries, each returning a single row with a single column so I don't know if to join them all in a single query will be a very good idea because we could impact server performance. At least there's not transaction lock.

Reply:
8 of 9 queries could be compressed in this:

```sql
SELECT 
    a.recompilations_sec, 
    b.user_connections,
    c.lock_waits_sec,
    d.page_splits_sec,
    e.compilations_sec,
    f.batch_req_sec,
    g.buffer_cache_hit_ratio,
    h.page_life_expectancy
FROM (
    SELECT cntr_value AS recompilations_sec
    FROM sys.dm_os_performance_counters 
    WHERE counter_name = 'SQL Re-Compilations/sec'
) AS a, (
    SELECT cntr_value AS user_connections
    FROM sys.dm_os_performance_counters 
    WHERE counter_name = 'User Connections'
) AS b, (
    SELECT cntr_value AS lock_waits_sec 
    FROM sys.dm_os_performance_counters 
    WHERE counter_name = 'Lock Waits/sec' 
        AND instance_name = '_Total'
) AS c, (
    SELECT cntr_value AS page_splits_sec 
    FROM sys.dm_os_performance_counters 
    WHERE counter_name = 'Page splits/sec'
) AS d, (
    SELECT cntr_value AS compilations_sec 
    FROM sys.dm_os_performance_counters 
    WHERE counter_name = 'SQL Compilations/sec'
) AS e, (
    SELECT cntr_value AS batch_req_sec 
    FROM sys.dm_os_performance_counters 
    WHERE counter_name = 'Batch Requests/sec'
) AS f, (
    SELECT (a.cntr_value * 1.0 / b.cntr_value) * 100.0 AS buffer_cache_hit_ratio 
    FROM sys.dm_os_performance_counters a 
    JOIN (
        SELECT cntr_value,OBJECT_NAME 
        FROM sys.dm_os_performance_counters 
        WHERE counter_name = 'Buffer cache hit ratio bASe' 
            AND OBJECT_NAME = 'SQLServer:Buffer Manager') b 
            ON  a.OBJECT_NAME = b.OBJECT_NAME 
                WHERE a.counter_name = 'Buffer cache hit ratio' 
                AND a.OBJECT_NAME = 'SQLServer:Buffer Manager'
) AS g, (
    SELECT [cntr_value] AS page_life_expectancy 
    FROM sys.dm_os_performance_counters 
    WHERE [object_name] = 'SQLServer:Buffer Manager' 
        AND [counter_name] = 'Page life expectancy'
) AS h
```

As far as I can see, times are similar.

Reply:
If you only get the last result in the data.json I think something is wrong in the metricset. Normally each `Fetch` of a metricset should emit 1 event (or in some metricsets multiple events of the same type). So far my understanding was that this Metricset should send one event?

I think the problem is here with our implementation of the Reporter. I had a look at the code again and it sends one event per query which I would think is not expected here? Instead it should build up 1 event out of all queries.

Reply:
Actually, it sends one event per row. It makes a lot of sense if you think about it. This way we can build up queries that return the same matrix of data which corresponds to our schema.

Reply:
I have changed the implementation and now it returns a matrix of results doing a single `SELECT` query that extracts the values it needs using a `WHERE` clause